include/seastar/coroutine/try_future.hh

```diff
@@ -0,0 +1,145 @@
+    auto hndl = std::coroutine_handle<U>::from_promise(*promise_ptr);
+
+    if (fut.failed()) {
+        hndl.promise().set_exception(std::move(fut).get_exception());
+        hndl.destroy();
+    } else {
+        hndl.resume();
+    }
+}
+
+template <bool CheckPreempt, typename T>
+class [[nodiscard]] try_future_awaiter : public seastar::task {
+    seastar::future<T> _future;
+    void (*_resume_or_destroy)(seastar::future<T>&, seastar::task&){};
+    seastar::task* _coroutine_task{};
+    seastar::task* _waiting_task{};
+
+public:
+    explicit try_future_awaiter(seastar::future<T>&& f) noexcept : _future(std::move(f)) {}
+
+    try_future_awaiter(const try_future_awaiter&) = delete;
+    try_future_awaiter(try_future_awaiter&&) = delete;
+
+    bool await_ready() const noexcept {
+        // Will suspend+schedule for ready failed futures too.
+        return _future.available() && !_future.failed() && (!CheckPreempt || !need_preempt());
+    }
+
+    template<typename U>
+    void await_suspend(std::coroutine_handle<U> hndl) noexcept {
+        _resume_or_destroy = try_future_resume_or_destroy_coroutine<T, U>;
+        _coroutine_task = &hndl.promise();
+        _waiting_task = hndl.promise().waiting_task();
+
+        if (!_future.available()) {
+            _future.set_coroutine(*this);
+        } else {
+            schedule(this);
+        }
+    }
+
+    T await_resume() {
+        if constexpr (std::is_void_v<T>) {
+            _future.get();
+        } else {
+            return std::move(_future).get();
+        }
+    }
+
+    virtual void run_and_dispose() noexcept override {
+        _resume_or_destroy(_future, *_coroutine_task);
+    }
+
+    virtual task* waiting_task() noexcept override {
+        return _waiting_task;
+    }
+};
+
+} // namespace seastar::internal
+
+namespace seastar::coroutine {
+
+/// \brief co_await:s a \ref future and returns the wrapped result if successful,
+/// terminates the coroutine otherwise, propagating the exception directly to the
+/// waiter.
+///
+/// If the future was successful, this is identical to co_await-ing the future
+/// directly. If the future failed, the coroutine is not resumed and instead the
+/// exception from the future is forwarded to the waiter directly and the
+/// coroutine is destroyed.
+///
+/// For example:
+/// ```
+/// // Function careful to not throw exceptions, instead returning failed futures.
+/// future<int> bar() {
+///    if (something_bad_happened) {
+///        return make_exception_future<>(std::runtime_error("error"));
+///    }
+///    return result;
+/// }
+///
+/// future<> foo() {
+///     auto result = co_await coroutine::try_future(bar());
+///     // This code is only executed if bar() returned a successful future.
+///     // Otherwise the exception is forwarded to the waiter future directly
+///     // and the coroutine is destroyed.
+///     check_result(result);
+/// }
```

Comment:
How would `foo()` look like if there were no `try_future()`? If I understand it correctly -- like this

```
future<> foo() {
    auto result = co_await bar();
    check_result(result);
}
```

and the result is the same -- if bar() resolves into exceptional future, the foo() is interrupted and results into exceptional future either, isn't it?